service/history/timer_queue_active_task_executor.go

```diff
@@ -432,12 +434,14 @@ func (t *timerQueueActiveTaskExecutor) executeActivityRetryTimerTask(
 		return err
 	}
 	if mutableState == nil || !mutableState.IsWorkflowExecutionRunning() {
-		return nil
+		release(nil) // release(nil) so mutable state is not unloaded from cache
+		return consts.ErrWorkflowExecutionNotFound
 	}
 
 	// generate activity task
 	activityInfo, ok := mutableState.GetActivityInfo(task.EventID)
 	if !ok || task.Attempt < activityInfo.Attempt || activityInfo.StartedEventId != common.EmptyEventID {
+		release(nil) // release(nil) so mutable state is not unloaded from cache
```

Comment:
the logging below uses activityInfo from mutable state, so we have to `release` after that.